src/pymor/algorithms/newton.py

```diff
@@ -124,6 +128,15 @@ def newton(operator, rhs, initial_guess=None, mu=None, error_norm=None, least_sq
             correction = jacobian.apply_inverse(residual, least_squares=least_squares)
         except InversionError:
             raise NewtonError('Could not invert jacobian')
+        if line_search_method is not None:
+            if line_search_method == 'armijo':
+                def res(x):
+                    residual_vec = rhs - operator.apply(x, mu=mu)
+                    return residual_vec.l2_norm()[0] if error_norm is None else error_norm(residual_vec)[0]
+                grad = 2.0 * jacobian.apply(residual) if error_norm is None else None
```

Comment:
Could you shortly explain to me how you compute this gradient, i.e. why there is this 2.0 *  ? Is it obvious?

Reply:
Here I'm taking the gradient of the **squared** l2-norm of the residual (one could also switch to the l2-norm itself). Consequently, the outer derivative should be 2*residual and the inner derivative is the jacobian of the residual. If I'm wrong with that, please let me know.

Reply:
Ah ok I see, I thought that the 2 from the derivative is already hidden in the jacobian of the operator but I am apparently not completely aware of the jacobian method. 
I think squared l2 norm is fine but in res(x) you are actually computing the norm... 
anyways, have you tested (for example by finite differences) that this is actually the correct gradient?
Is there no chance of getting the gradient in case error norm is not `None` ?

Reply:
What do you think about dropping `error_norm` in favor of `error_product`? Then `grad` could be computed. I don't know anybody who uses pyMOR with non-hilbert norms, and if your norms come from a product `Operator`, it's easier to directly pass that operator instead of having to pass `induced_norm(product)`.

Reply:
I thought a bit more about the gradient and I think it is not correct the way it is at the moment. IMHO, there is a jacobian transposed missing.

@sdrave Sounds good. I think switching to `error_product` would affect some other parts in pyMOR too, right?

Reply:
Actually, the `error_norm` parameter of `newton` is used nowhere in all of pyMOR ..

Reply:
I switched to `error_product` in  f661730. Maybe you can have a look? I'm still not sure about the gradient, though.

I do not take the root after applying the `error_product` as to not loose any precision. When it comes down to convergence or not, this should not make a big difference; but certainly, the error bounds are not the same anymore.

Reply:
This is good. 
I think an easy test for the gradient is to check it with simple finite differences. Then you get an idea whether this is the correct one. Anyways, where did you get the plus from?

Reply:
Do you mean the plus here
```
- (jacobian.apply(residual) + jacobian.apply_adjoint(residual))
```
or somewhere else? I will try whether or not finite differences lead to the same result.

And I will add line breaks of course, sorry for missing that.

Reply:
To be extra sure, checking numerically would be good, but I am quite confident, that the formula is correct.

Reply:
After a rough numerical check using finite differences, I'm also confident that the formula is working as expected.